# Volatility

The _Volatility_ of a stock gives us a measure on the amount of uncertainty given the changes made over time. Normally, it is computed using the standard deviation over the logaritmic returns.

The daily volatility for a given period of returns $P$ can be defined as:

$\displaystyle \sigma _{\text{P}}=\sigma _{\text{daily}}{\sqrt{P}}$

Therefore, to calculate the volatily for one month, we just have to compute the standard deviation over the returns of that month, having $P$ trading days:

In [1]:
from math import sqrt
from sample_data import stock_prices

msft = stock_prices("msft")

time_slice = msft.Close['2016-03-30':'2016-04-29']
std_dev_returns = time_slice.pct_change().std()
trading_days = len(time_slice)

volatility = std_dev_returns * sqrt(trading_days)
volatility

np.float64(0.08540732054646473)

This gives us notion on how far the price might deviate from the average, in this case a 8.54%.

## Beta

Beta (β) is a measure of the volatility in relation to the market as a whole. It's normally computed against an index like SP&500. Stocks having β values below 0 indicate that its returns move close to the index. Stocks with betas higher than 1.0 can be interpreted as more volatile than the S&P 500.

It's defined as:

$\displaystyle \beta _{i}={\frac {\mathrm {Cov} (r_{i},r_{m})}{\mathrm {Var} (r_{m})}}$

where $r_{i}$ is the return on an individual stock and $r_{m}$ is the return on the overall market.

To calculate it, first we need to get the data for the S&P500 index, represented as `^GSPC`. We will use a library called [pandas_datareader](https://pydata.github.io/pandas-datareader/remote_data.html), that can read stock data from different sources. Alternativelly, you can read the csv file in `./data/sp500.csv`.

In [2]:
import pandas as pd
sp500 = pd.read_csv('./data/sp500.csv', index_col=0, parse_dates=[0])
sp500[:3]

# Alternativelly
# import pandas_datareader.data as dr
# sp500 = dr.DataReader("^GSPC", "yahoo", start='2018-01-01', end='2022-12-31')

High          Low         Open        Close      Volume  \
Date                                                                         
2010-01-04  1133.869995  1116.560059  1116.560059  1132.989990  3991400000   
2010-01-05  1136.630005  1129.660034  1132.660034  1136.520020  2491020000   
2010-01-06  1139.189941  1133.949951  1135.709961  1137.140015  4972660000   

              Adj Close  
Date                     
2010-01-04  1132.989990  
2010-01-05  1136.520020  
2010-01-06  1137.140015

We will compute the Beta value for 5Y period:

In [3]:
aapl = stock_prices("aapl")
r_i = aapl.Close['2018':'2022'].pct_change()
r_m = sp500.Close['2018':'2022'].pct_change()
beta = r_i.cov(r_m) / r_m.var()
beta

np.float64(1.2530339335561118)